From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: noeppi_noeppi <noeppinoeppi@gmail.com>
Date: Sun, 5 May 2024 12:24:22 +0200
Subject: [PATCH] Add log files to .gitignore


diff --git a/.gitignore b/.gitignore
index 6856832a44711f7b00280e2fca000196681f6ede..45cb8b456ef1eb4d5d8dac5ed3f09e37da1a990b 100644
--- a/.gitignore
+++ b/.gitignore
@@ -137,3 +137,5 @@ log.txt
 
 # Fleet
 .fleet/
+
+log_*.txt.gz
