From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: noeppi_noeppi <noeppinoeppi@gmail.com>
Date: Sun, 5 May 2024 12:24:22 +0200
Subject: [PATCH] Add log files to .gitignore


diff --git a/.gitignore b/.gitignore
index be2837373a96597826bd168509b8e47686130405..2daa044cb1509c22f45f2e6e824ea461a6b383a2 100644
--- a/.gitignore
+++ b/.gitignore
@@ -136,3 +136,5 @@ log.txt
 
 # Fleet
 .fleet/
+
+log_*.txt.gz
